common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -117,76 +118,490 @@ public Collation(
+              throw new SparkException("UNSPECIFIED collation specifier reserved for internal use");
+            } else if (Arrays.stream(CaseSensitivity.values()).anyMatch(
+                (s) -> s.toString().equals(part))) {
+              specifiers |=
+                CaseSensitivity.valueOf(part).ordinal() << caseSensitivityOffset;
+            } else if (Arrays.stream(AccentSensitivity.values()).anyMatch(
+                (s) -> part.equals(s.toString()))) {
+              specifiers |=
+                AccentSensitivity.valueOf(part).ordinal() << accentSensitivityOffset;
+            } else if (Arrays.stream(PunctuationSensitivity.values()).anyMatch(
+                (s) -> part.equals(s.toString()))) {
+              specifiers |=
+                PunctuationSensitivity.valueOf(part).ordinal() << punctuationSensitivityOffset;
+            } else if (Arrays.stream(FirstLetterPreference.values()).anyMatch(
+                (s) -> part.equals(s.toString()))) {
+              specifiers |=
+                FirstLetterPreference.valueOf(part).ordinal() << firstLetterPreferenceOffset;
+            } else if (Arrays.stream(CaseConversion.values()).anyMatch(
+                (s) -> part.equals(s.toString()))) {
+              specifiers |=
+                CaseConversion.valueOf(part).ordinal() << caseConversionOffset;
+            } else if (Arrays.stream(SpaceTrimming.values()).anyMatch(
+                (s) -> part.equals(s.toString()))) {
+              specifiers |=
+                SpaceTrimming.valueOf(part).ordinal() << spaceTrimmingOffset;
+            } else {
+              throw new SparkException("Invalid collation specifier value " + part);
+            }
+          }
+        }
+        return specifiers;
+      }
+
+      public Collation buildCollation() {
+        if (implementationProvider == ImplementationProvider.UTF8_BINARY) {
+          return buildUTF8BinaryCollation();
+        } else {
+          return buildICUCollation();
+        }
+      }
+
+      public Collation buildUTF8BinaryCollation() {
+        Comparator<UTF8String> comparator;
+        if (collationId == UTF8_BINARY_COLLATION_ID) {
+          comparator = UTF8String::binaryCompare;
+        } else {
+          comparator = (s1, s2) -> {
+            UTF8String convertedS1 = caseAndTrimmingConversionUTF8Binary(s1);
+            UTF8String convertedS2 = caseAndTrimmingConversionUTF8Binary(s2);
+            return convertedS1.binaryCompare(convertedS2);
+          };
+        }
+        return new Collation(
+          collationName(),
+          null,
+          comparator,
+          "1.0",
+          s -> (long) caseAndTrimmingConversionUTF8Binary(s).hashCode(),
+          collationId == UTF8_BINARY_COLLATION_ID,
+          collationId == UTF8_BINARY_COLLATION_ID,
+          collationId == UTF8_BINARY_LCASE_COLLATION_ID
+        );
+      }
+
+      private UTF8String caseAndTrimmingConversionUTF8Binary(UTF8String s) {
```

Comment:
My suggestion would be to only cover collation id format, as you did above. No need to implement support/parsing for anything but case and accent at this point.
We can follow up with trimming and co support.